### ISSUE TITLE
PyTorch is not linked with support for mps devices

### ISSUE THREAD

**Issue by Aaack-Aardvark** [yukiarimo/yuna-ai#108]
After switching to the Dev branch I get the error:

`"RuntimeError: PyTorch is not linked with support for mps devices"`

The full error is 

```
Traceback (most recent call last):
  File "/home/spike/Documents/Yuna/yuna-ai/index.py", line 4, in <module>
    from lib.generate import ChatGenerator, ChatHistoryManager, get_config
  File "/home/spike/Documents/Yuna/yuna-ai/lib/generate.py", line 14, in <module>
    from lib import router
  File "/home/spike/Documents/Yuna/yuna-ai/lib/router.py", line 5, in <module>
    from lib.vision import capture_image
  File "/home/spike/Documents/Yuna/yuna-ai/lib/vision.py", line 2, in <module>
    from lib.audio import speak_text
  File "/home/spike/Documents/Yuna/yuna-ai/lib/audio.py", line 13, in <module>
    yunaListen = pipeline(
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/pipelines/__init__.py", line 1102, in pipeline
    return pipeline_class(model=model, framework=framework, task=task, **kwargs)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/pipelines/automatic_speech_recognition.py", line 219, in __init__
    super().__init__(model, tokenizer, feature_extractor, device=device, torch_dtype=torch_dtype, **kwargs)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/pipelines/base.py", line 901, in __init__
    self.model.to(self.device)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/modeling_utils.py", line 2958, in to
    return super().to(*args, **kwargs)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1174, in to
    return self._apply(convert)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 780, in _apply
    module._apply(fn)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 780, in _apply
    module._apply(fn)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 780, in _apply
    module._apply(fn)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 805, in _apply
    param_applied = fn(param)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1160, in convert
    return t.to(
RuntimeError: PyTorch is not linked with support for mps devices
```

To reproduce just git checkout Dev, git pull and try to start Yuna with python3 index.py

I don't know which other device can I set put in the `/static/config,json`

I'n on linux PopOS 22.04, using venv.  I can't use anaconda sadly.  If I trace the files of the error and just comment out whatever refers to audio, Yuna starts, also the main branch works as expected, I guess mps is a device that only runs on MacOS.

**Comment by yukiarimo** [comment]
Of course, because you’re on Linux! MPS is an Apple Silicon stuff. Please change it to CUDA, and make sure all staff is installed correctly with the necessary support (ROCM will be used for AMD, but still should be stated as CUDA)

**Comment by yukiarimo** [comment]
Let me know if it'll work! Also, V7.0 is coming very soon, so keep updated!

**Comment by Aaack-Aardvark** [comment]
Wow, thank you for answering so fast and in a sunday!
Ok I got the same error, changed "mps" to "CUDA" in line 42 of the /static/config,.json file ("device": "CUDA",)
Now the error changed:

Traceback (most recent call last):
  File "/home/spike/Documents/Yuna/yuna-ai/index.py", line 4, in <module>
    from lib.generate import ChatGenerator, ChatHistoryManager, get_config
  File "/home/spike/Documents/Yuna/yuna-ai/lib/generate.py", line 14, in <module>
    from lib import router
  File "/home/spike/Documents/Yuna/yuna-ai/lib/router.py", line 5, in <module>
    from lib.vision import capture_image
  File "/home/spike/Documents/Yuna/yuna-ai/lib/vision.py", line 2, in <module>
    from lib.audio import speak_text
  File "/home/spike/Documents/Yuna/yuna-ai/lib/audio.py", line 13, in <module>
    yunaListen = pipeline(
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/pipelines/__init__.py", line 1102, in pipeline
    return pipeline_class(model=model, framework=framework, task=task, **kwargs)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/pipelines/automatic_speech_recognition.py", line 219, in __init__
    super().__init__(model, tokenizer, feature_extractor, device=device, torch_dtype=torch_dtype, **kwargs)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/pipelines/base.py", line 901, in __init__
    self.model.to(self.device)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/modeling_utils.py", line 2958, in to
    return super().to(*args, **kwargs)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1174, in to
    return self._apply(convert)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 780, in _apply
    module._apply(fn)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 780, in _apply
    module._apply(fn)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 780, in _apply
    module._apply(fn)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 805, in _apply
    param_applied = fn(param)
  File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1160, in convert
    return t.to(
RuntimeError: PyTorch is not linked with support for mps devices

It complains about the same in different lines.  I'm sorry if this is a noob error, I barely know how to edit a text file so python is ancient magic to me >_<!

**Comment by cloner174** [comment]
> Wow, thank you for answering so fast and in a sunday! Ok I got the same error, changed "mps" to "CUDA" in line 42 of the /static/config,.json file ("device": "CUDA",) Now the error changed:
> 
> Traceback (most recent call last): File "/home/spike/Documents/Yuna/yuna-ai/index.py", line 4, in from lib.generate import ChatGenerator, ChatHistoryManager, get_config File "/home/spike/Documents/Yuna/yuna-ai/lib/generate.py", line 14, in from lib import router File "/home/spike/Documents/Yuna/yuna-ai/lib/router.py", line 5, in from lib.vision import capture_image File "/home/spike/Documents/Yuna/yuna-ai/lib/vision.py", line 2, in from lib.audio import speak_text File "/home/spike/Documents/Yuna/yuna-ai/lib/audio.py", line 13, in yunaListen = pipeline( File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/pipelines/**init**.py", line 1102, in pipeline return pipeline_class(model=model, framework=framework, task=task, **kwargs) File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/pipelines/automatic_speech_recognition.py", line 219, in **init** super().**init**(model, tokenizer, feature_extractor, device=device, torch_dtype=torch_dtype, **kwargs) File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/pipelines/base.py", line 901, in **init** self.model.to(self.device) File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/transformers/modeling_utils.py", line 2958, in to return super().to(*args, **kwargs) File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1174, in to return self._apply(convert) File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 780, in _apply module._apply(fn) File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 780, in _apply module._apply(fn) File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 780, in _apply module._apply(fn) File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 805, in _apply param_applied = fn(param) File "/home/spike/Documents/Yuna/Yuna/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1160, in convert return t.to( RuntimeError: PyTorch is not linked with support for mps devices
> 
> It complains about the same in different lines. I'm sorry if this is a noob error, I barely know how to edit a text file so python is ancient magic to me >_<!

--------


Hi, I hope you are doing well. Since Yuki is unavailable for a few days, I’m responding on his behalf. It seems there is an issue where the device setting isn't being applied correctly. Please go to the file ```yuna-ai/lib/audio.py``` and modify the following line:

```
yunaListen = whisper.load_model(name="tiny.en", device="cpu", in_memory=True)
```

to:

```
yunaListen = whisper.load_model(name="tiny.en", device="cuda" if torch.cuda.is_available() else "cpu")
```

This will manually set the device.

**Comment by Aaack-Aardvark** [comment]
Thank you very much for the help, I'll try it and report tomorrow! Very much appreciated.

**Comment by Aaack-Aardvark** [comment]
ok new error, i checked out dev, git pull to make sure everything was up to date and edited lib/audio.py.
The line that was closest to the example was #13 it contained this:

yunaListen = pipeline(
    "automatic-speech-recognition",
    model="openai/whisper-tiny",
    torch_dtype=torch.float32,
    device="mps",  # or mps for Mac devices
    model_kwargs={"attn_implementation": "sdpa"},
)

so I commented it out and replace it with the line you pasted and i got this error:

Traceback (most recent call last):
  File "/home/spike/Documents/Yuna/yuna-ai/index.py", line 4, in <module>
    from lib.generate import ChatGenerator, ChatHistoryManager, get_config
  File "/home/spike/Documents/Yuna/yuna-ai/lib/generate.py", line 14, in <module>
    from lib import router
  File "/home/spike/Documents/Yuna/yuna-ai/lib/router.py", line 5, in <module>
    from lib.vision import capture_image
  File "/home/spike/Documents/Yuna/yuna-ai/lib/vision.py", line 2, in <module>
    from lib.audio import speak_text
  File "/home/spike/Documents/Yuna/yuna-ai/lib/audio.py", line 21, in <module>
    yunaListen = whisper.load_model(
NameError: name 'whisper' is not defined

so I uncommented the original line and changed the "mps" part with "CUDA" instead, new error!

RuntimeError: Expected one of cpu, cuda, ipu, xpu, mkldnn, opengl, opencl, ideep, hip, ve, fpga, maia, xla, lazy, vulkan, mps, meta, hpu, mtia, privateuseone device type at start of device string: CUDA

Ooops it turns out "CUDA" should be lowercase corrected it on audio.py and config.json and...

ValueError: Model path does not exist: lib/models/yuna/fused_q6k.gguf

ah ok, I need to change the model, I changed to yuna-ai-v3-atomic-q_6_k.gguf which is the one I downloaded and that did the trick! It started!!
I'm not sure if the audio works but now I have a working dev branch copy of Yuna-ai

so to summarize i did the following changes:

static/config.json:

"yuna_default_model": "yuna-ai-v3-atomic-q_6_k.gguf",
and 
"device": "cuda",

lib/audio.py:

yunaListen = pipeline(
    "automatic-speech-recognition",
    model="openai/whisper-tiny",
    torch_dtype=torch.float32,
    device="cuda",  # or mps for Mac devices
    model_kwargs={"attn_implementation": "sdpa"},
)

**Comment by Aaack-Aardvark** [comment]
I forgot to ask, should I close this issue?

**Comment by yukiarimo** [comment]
Sure! Yeah, guys just wait, there some issue with Kanojo, I’m working on that! Closing for now as resolved!
